openstates/al/bills.py

```diff
@@ -39,222 +43,357 @@ def _categorize_action(action):
-            if not title:
-                return
-
-            # create bill
-            bill = Bill(session, chamber, bill_id, title, type=bill_type)
+            else:
+                raise AssertionError(
+                        "Unknown bill type for bill '{}'".format(bill_id))
+
+            bill = Bill(
+                    session=self.session,
+                    chamber=chamber,
+                    bill_id=bill_id,
+                    title='',
+                    type=bill_type
+                    )
             if subject:
                 bill['subjects'] = [subject]
+            if sponsor:
+                bill.add_sponsor(type='primary', name=sponsor)
+            bill.add_source(url)
+
+            bill_url = 'http://alisondb.legislature.state.al.us/Alison/SESSBillResult.aspx?BILL={}'.format(bill_id)
+            bill.add_source(bill_url)
+            try:
+                bill_html = self.sget(url=bill_url).text
+            except AssertionError:
+                self.warning("The webpage for bill {} isn't loading, so it won't be saved".format(bill_id))
+                continue
+            bill_doc = lxml.html.fromstring(bill_html)
 
-            if fnotes == 'ON':
-                bill.add_document('fiscal notes', 'http://alisondb.legislature.state.al.us/acas/ACTIONFiscalNotesFrameMac.asp?OID=%s&LABEL=%s' %
-                                  (oid, bill_id))
-
-            self.get_sponsors(bill, oid)
-            self.get_actions(bill, oid)
+            title = bill_doc.xpath('//div[@class="shorttitle"]')[0].text_content().strip()
+            if not title:
+                title = "[No title given by state]"
+            bill['title'] = title
+
+            (viewstate, ) = bill_doc.xpath('//input[@id="__VIEWSTATE"]/@value')
+            (viewstategenerator, ) = bill_doc.xpath('//input[@id="__VIEWSTATEGENERATOR"]/@value')
+            form = {
+                    '__EVENTTARGET': '',
+                    '__EVENTARGUMENT': '',
+                    '__LASTFOCUS': '',
+                    '__VIEWSTATE': viewstate,
+                    '__VIEWSTATEGENERATOR': viewstategenerator,
+                    'ctl00$cboSession': self.session_name,
+                    }
+
+            version_url_base = 'http://alisondb.legislature.state.al.us/ALISON/SearchableInstruments/{0}/PrintFiles/{1}-'.format(self.session, bill_id)
+            versions = bill_doc.xpath('//table[@class="box_versions"]/tr/td[2]/font/text()')
+            for version in versions:
+                name = version
+                if version == "Introduced":
+                    version_url = version_url_base + 'int.pdf'
+                elif version == "Engrossed":
+                    version_url = version_url_base + 'eng.pdf'
+                elif version == "Enrolled":
+                    version_url = version_url_base + 'enr.pdf'
+                else:
+                    raise NotImplementedError("Unknown version type found: '{}'".format(name))
```

Comment:
Yes, they're fetched using a complex `postback`. Fortunately accessing their URLs are ALISON session agnostic, so they'll work for anyone, without cookies.